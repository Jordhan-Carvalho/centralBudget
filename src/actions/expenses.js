import db from '../firebase/firebase';


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
  });

    
  export const startAddExpense = (expenseData = {}) => {
    // only works bc redux middleware (redux thunk)
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0,
      } = expenseData;
      const expense = {description, note, amount, createdAt};
      //save the data to firebase (aynsc call) (used the return to testing porpouse)
        const ref = await db.ref(`users/${uid}/expenses`).push(expense);
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    };
  };
  
  // REMOVE_EXPENSE
  export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
  });

  export const startRemoveExpense = (expense) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
        const id = expense.id;
  try{
  await db.ref(`users/${uid}/expenses/${id}`).remove();
        dispatch(removeExpense(expense));
      } catch (error) {
        alert(error);
      }
      };
      };
  
  // EDIT_EXPENSE
  export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
  });

  export const startEditExpense = (id, updates) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      try{
        await db.ref(`users/${uid}/expenses/${id}`).update(updates);
        dispatch(editExpense(id, updates));
      } catch (error) {
        console.log(error);
      }
    };
  };

  // SET_EXPENSES
  export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
  });

  // export const startSetExpenses
  export const startSetExpenses = () => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
  // 1. fetch all expense data once
  //2. Parse the data to array
  const expenses = []; 
  await db.ref(`users/${uid}/expenses`).once('value', (snapshot) => {
        snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
        });
        // 3. dispatch SET_EXPENSES
        dispatch(setExpenses(expenses));
      };
      };