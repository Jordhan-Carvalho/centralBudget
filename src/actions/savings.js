import db from '../firebase/firebase';

// ADD_SAVING
export const addSaving = (saving) => ({
    type: 'ADD_SAVING',
    saving
  });

    
  export const startAddSaving = (savingData = {}) => {
    // only works bc redux middleware (redux thunk)
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
      } = savingData;
      const saving = {description, note, amount, createdAt};
      //save the data to firebase (aynsc call) (used the return to testing porpouse)
        const ref = await db.ref(`users/${uid}/savings`).push(saving);
      dispatch(addSaving({
        id: ref.key,
        ...saving
      }));
    };
  };
  
  // REMOVE_EXPENSE
  export const removeSaving = ({ id } = {}) => ({
    type: 'REMOVE_SAVING',
    id
  });

  export const startRemoveSaving = (saving) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
        const id = saving.id;
  try{
  await db.ref(`users/${uid}/savings/${id}`).remove();
        dispatch(removeSaving(saving));
      } catch (error) {
        alert(error);
      }
      };
      };
  
  // EDIT_EXPENSE
  export const editSaving = (id, updates) => ({
    type: 'EDIT_SAVING',
    id,
    updates
  });

  export const startEditSaving = (id, updates) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      try{
        await db.ref(`users/${uid}/savings/${id}`).update(updates);
        dispatch(editSaving(id, updates));
      } catch (error) {
        console.log(error);
      }
    };
  };

  // SET_EXPENSES
  export const setSavings = (savings) => ({
    type: 'SET_SAVINGS',
    savings
  });

  // export const startSetExpenses
  export const startSetSavings = () => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
  // 1. fetch all saving data once
  //2. Parse the data to array
  const savings = []; 
  await db.ref(`users/${uid}/savings`).once('value', (snapshot) => {
        snapshot.forEach((childSnapshot) => {
                savings.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
        });
        // 3. dispatch SET_EXPENSES
        dispatch(setSavings(savings));
      };
      };