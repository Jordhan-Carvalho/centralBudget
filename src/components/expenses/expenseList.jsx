import React from "react";
import { connect } from "react-redux";
import ListItem from "../common/listItem";
import selectVisible from "../../selectors/visibleTrans";

export const ExpenseList = props => {
  return (
    <div>
      {props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        <ListItem type={"expense"} expenses={props.expenses} />
      )}
    </div>
  );
};

// redux call
const mapStateToProps = state => {
  return {
    expenses: selectVisible(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
