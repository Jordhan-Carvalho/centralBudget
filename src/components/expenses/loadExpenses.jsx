import React, { Component } from "react";
import { Link } from "react-router-dom";
import ExpenseList from "./expenseList";
import ListFilters from "../common/listFilters";
import TransSumary from "../common/transSumary";

class LoadExpenses extends Component {
  state = {};
  render() {
    return (
      <React.Fragment>
        <h1>Expense List</h1>
        <hr />
        <ListFilters />
        <TransSumary type={"expense"} />
        <hr />
        <ExpenseList />
        <hr />
        <Link to="/create">
          <button className="btn btn-dark">Add Expense</button>
        </Link>
      </React.Fragment>
    );
  }
}

export default LoadExpenses;
