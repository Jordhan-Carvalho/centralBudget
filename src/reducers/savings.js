const savingsReducerDefaultState = [];

const savingsReducer = (state = savingsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_SAVING':
      return [
        ...state,
        action.saving
      ];
    case 'REMOVE_SAVING':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_SAVING':
      return state.map((saving) => {
        if (saving.id === action.id) {
          return {
            ...saving,
            ...action.updates
          };
        } else {
          return saving;
        };
      });
      case 'SET_SAVINGS':
      return action.savings;
    default:
      return state;
  }
};


export default savingsReducer;