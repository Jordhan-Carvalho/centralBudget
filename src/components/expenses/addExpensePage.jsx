import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "../common/form";
import { startAddExpense } from "../../actions/expenses";

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/expenses");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add Expense</h1>
        <Form onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
