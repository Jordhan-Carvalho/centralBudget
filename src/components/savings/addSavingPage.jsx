import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "../common/form";
import { startAddSaving } from "../../actions/savings";

export class AddSavingPage extends Component {
  onSubmit = saving => {
    this.props.startAddSaving(saving);
    this.props.history.push("/savings");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add Saving</h1>
        <Form onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    startAddSaving: saving => dispatch(startAddSaving(saving))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddSavingPage);
