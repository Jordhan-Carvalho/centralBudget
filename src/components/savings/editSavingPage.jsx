import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "../common/form";
import { startEditSaving, startRemoveSaving } from "../../actions/savings";

export class EditSavingPage extends Component {
  onRemove = () => {
    this.props.startRemoveSaving(this.props.saving);
    this.props.history.push("/savings");
  };

  onSubmit = ({ description, note, amount, createdAt }) => {
    const saving = { description, note, amount, createdAt };
    this.props.startEditSaving(this.props.saving.id, saving);
    this.props.history.push("/savings");
  };

  render() {
    return (
      <div className="form-group">
        <h1>Edit {this.props.saving.description}</h1>
        <div className="form-group">
          <Form
            expense={this.props.saving}
            onSubmit={this.onSubmit}
            isIncome={false}
          />
        </div>
        <div className="form-group">
          <button className="btn btn-danger" onClick={this.onRemove}>
            Remove
          </button>
        </div>
      </div>
    );
  }
}

// maps redux

const mapDispatchToProps = dispatch => {
  return {
    startEditSaving: (savingID, saving) =>
      dispatch(startEditSaving(savingID, saving)),
    startRemoveSaving: saving => dispatch(startRemoveSaving(saving))
  };
};

const mapStateToProps = (state, props) => {
  return {
    saving: state.savings.find(saving => {
      return saving.id === props.match.params.id;
    })
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditSavingPage);
