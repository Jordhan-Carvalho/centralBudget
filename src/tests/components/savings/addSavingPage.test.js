import React from 'react';
import { shallow } from 'enzyme';
import { AddSavingPage } from '../../../components/savings/addSavingPage';
import expenses from '../../fixtures/expenses';

// using lifecycle methods (global) refactoring
let startAddSaving, history, wrapper;
beforeEach(() => {
    startAddSaving = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddSavingPage startAddSaving={startAddSaving} history={history} />);
});
//

test('should render AddSavingPage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

//test in case of submiting everything right
test('should handle onSubmit', () => {
    wrapper.find('Form').prop('onSubmit')(expenses[0]);
    //check if spies (mock fn) are ij
    expect(history.push).toHaveBeenLastCalledWith('/savings');
    expect(startAddSaving).toHaveBeenLastCalledWith(expenses[0]);
});