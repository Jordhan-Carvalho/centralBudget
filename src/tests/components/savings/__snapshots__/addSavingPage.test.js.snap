// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddSavingPage correctly 1`] = `
<Fragment>
  <h1>
    Add Saving
  </h1>
  <Form
    onSubmit={[Function]}
  />
</Fragment>
`;
